### PR TITLE
feat: flake8-use-pathlib PTH100-124

### PR DESCRIPTION
**Pull request by sbrugman**
See: https://github.com/charliermarsh/ruff/issues/2060

For review: The checking if rules are enabled, and the mapping from call path => violation could probably be simplified, but I don't yet see how - suggestions welcome.

### CODE REVIEW

File: src/checkers/ast.rs [hunk 1]
```diff
@@ -2545,6 +2545,33 @@ where
...
+                    || self.settings.rules.enabled(&Rule::PathlibIsDir)
+                    || self.settings.rules.enabled(&Rule::PathlibIsFile)
+                    || self.settings.rules.enabled(&Rule::PathlibIsLink)
+                    || self.settings.rules.enabled(&Rule::PathlibReadlink)
+                    || self.settings.rules.enabled(&Rule::PathlibStat)
+                    || self.settings.rules.enabled(&Rule::PathlibIsAbs)
+                    || self.settings.rules.enabled(&Rule::PathlibJoin)
+                    || self.settings.rules.enabled(&Rule::PathlibBasename)
+                    || self.settings.rules.enabled(&Rule::PathlibSamefile)
+                    || self.settings.rules.enabled(&Rule::PathlibSplitext)
```
Comment (charliermarsh) [inline src/checkers/ast.rs:2571]: This is basically the pattern we use elsewhere. It's not great, but no major suggestions on how to change it right now, so all good.

File: src/rules/flake8_use_pathlib/helpers.rs [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+                OsCall::Stat => PathlibStat.into(),
+                OsCall::IsAbs => PathlibIsAbs.into(),
+                OsCall::Join => PathlibJoin.into(),
+                OsCall::Basename => PathlibBasename.into(),
+                OsCall::Dirname => PathlibDirname.into(),
+                OsCall::Samefile => PathlibSamefile.into(),
+                OsCall::Splitext => PathlibSplitext.into(),
+            },
+            Range::from_located(expr),
+        );
```
Comment (charliermarsh) [inline src/rules/flake8_use_pathlib/helpers.rs:235]: Is it possible to get rid of all the `if checker.settings.rules.enabled` in the above `map` call, then here, create the `Diagnostic`, and do `settings.rules.enabled(diagnostic.kind.rule())` prior to pushing?

Comment (sbrugman) [reply src/rules/flake8_use_pathlib/helpers.rs:97]: Makes sense, pushed the changes. We already checked that any of them is enabled. We might even be able to get rid of `OsCall` altogether

File: src/rules/flake8_use_pathlib/violations.rs [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+}
+
+// PTH101
+define_violation!(
+    pub struct PathlibChmod;
+);
+impl Violation for PathlibChmod {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("`os.chmod` should be replaced by `.chmod`")
```
Comment (charliermarsh) [inline src/rules/flake8_use_pathlib/violations.rs:24]: Should this be `.chmod()` at the end, with parens?

File: src/rules/flake8_use_pathlib/violations.rs [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+}
+
+// PTH116
+define_violation!(
+    pub struct PathlibStat;
+);
+impl Violation for PathlibStat {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("`os.stat` should be replaced by `.stat()` or `.owner()` or `.group`")
```
Comment (charliermarsh) [inline src/rules/flake8_use_pathlib/violations.rs:189]: Is `.group` missing parens?

File: src/rules/flake8_use_pathlib/helpers.rs (hunk 2 above)
Comment (charliermarsh) [reply src/rules/flake8_use_pathlib/helpers.rs:97]: Yeah you could probably get rid of `OsCall` here?

File: src/rules/flake8_use_pathlib/violations.rs (hunk 3 above)
Comment (charliermarsh) [reply src/rules/flake8_use_pathlib/violations.rs:24]: (Genuinely asking, since I know some of these are actually properties, like `.parent`.)

Comment (sbrugman) [reply src/rules/flake8_use_pathlib/violations.rs:24]: Yes, was a mistake

File: src/rules/flake8_use_pathlib/violations.rs (hunk 4 above)
Comment (sbrugman) [reply src/rules/flake8_use_pathlib/violations.rs]: Fixed

File: src/rules/flake8_use_pathlib/helpers.rs (hunk 2 above)
Comment (sbrugman) [reply src/rules/flake8_use_pathlib/helpers.rs:104]: Fixed in https://github.com/charliermarsh/ruff/pull/2091

**Pull request merged**